pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -1305,3 +1333,67 @@ func (testCases ListUsersTests) runListUsersTestCases(t *testing.T) {
 		})
 	}
 }
+
+func TestListUsersCycleDetection(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+	storeID := ulid.Make().String()
+	modelID := ulid.Make().String()
+
+	mockController := gomock.NewController(t)
+	defer mockController.Finish()
+
+	mockDatastore := mocks.NewMockOpenFGADatastore(mockController)
+
+	// Times(0) ensures that we exit quickly
+	mockDatastore.EXPECT().Read(gomock.Any(), gomock.Any(), gomock.Any()).Times(0)
+
+	l := NewListUsersQuery(mockDatastore)
+	channelWithResults := make(chan *openfgav1.User)
+	model := testutils.MustTransformDSLToProtoWithID(`
+	model
+		schema 1.1
+	type user
+	type document
+		relations
+			define viewer: [user]
+	`)
+	typesys := typesystem.New(model)
+	ctx := typesystem.ContextWithTypesystem(context.Background(), typesys)
+
+	t.Run("enters_loop_detection", func(t *testing.T) {
+		visitedUserset := &openfgav1.UsersetUser{
+			Type:     "document",
+			Id:       "1",
+			Relation: "viewer",
+		}
+		visitedUsersetKey := fmt.Sprintf("%s:%s#%s", visitedUserset.GetType(), visitedUserset.GetId(), visitedUserset.GetRelation())
+		visitedUsersets := make(map[string]struct{})
+		visitedUsersets[visitedUsersetKey] = struct{}{}
+
+		err := l.expand(ctx, &internalListUsersRequest{
+			ListUsersRequest: &openfgav1.ListUsersRequest{
+				StoreId:              storeID,
+				AuthorizationModelId: modelID,
+				Object: &openfgav1.Object{
+					Type: visitedUserset.GetType(),
+					Id:   visitedUserset.GetId(),
+				},
+				Relation: visitedUserset.GetRelation(),
+				UserFilters: []*openfgav1.ListUsersFilter{{
+					Type: "user",
+				}},
+			},
+			visitedUsersetsMap: visitedUsersets,
+		}, channelWithResults)
+
+		require.NoError(t, err)
+		select {
+		case <-channelWithResults:
```

Comment:
question: do we need go func to wrap around the expand call?  It feels weird to see us listening on a channel but the originator is not a go routine.

Reply:
good point!